### PR TITLE
Calendar redesign introduction

### PR DESCRIPTION
**Pull request by BaqWin**
Calendar redesign introduction

### CODE REVIEW

File: src/app/calendar/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,4 @@
+add_library(calendar-lib calendar.cpp connector.cpp)
+
+target_compile_options(calendar-lib PRIVATE -Werror -Wall -Wextra -pedantic)
```
Comment (fbtom) [inline src/app/calendar/CMakeLists.txt:3]: You can add a 
target_link_libraries(clinic-lib PUBLIC doctor-lib rom-lib)
to remove `"../clinic/"` in calendar.cpp

File: src/app/calendar/calendar.cpp [hunk 2]
```diff
@@ -0,0 +1,27 @@
+#include "calendar.hpp"
+#include "../clinic/clinic_facade.hpp"
```
Comment (fbtom) [inline src/app/calendar/calendar.cpp:2]: You can remove `"../clinic/"`

File: src/app/clinic/clinic_facade.hpp [hunk 3]
```diff
@@ -1,4 +1,5 @@
 #pragma once
+#include "../calendar/calendar.hpp"
```
Comment (fbtom) [inline src/app/clinic/clinic_facade.hpp:2]: Adjust CMake and remove `"../x"`, same as #99

File: src/app/calendar/calendar.hpp [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+#include "../room/room.hpp"
+#include "connector.hpp"
+#include <chrono>
+#include <map>
+#include <memory>
+#include <stdexcept>
+
+class Calendar
+{
+    std::multimap<std::chrono::year_month_day, std::shared_ptr<Connector>> map_calendar_;
```
Comment (fbtom) [inline src/app/calendar/calendar.hpp:12]: With a container memory is already on heap, so you can keep it bit simpler by multimap of ymd and connectors

File: src/app/calendar/connector.cpp [hunk 5]
```diff
@@ -0,0 +1,10 @@
+#include "connector.hpp"
+
+Connector::Connector(const std::chrono::year_month_day& ymd, const std::shared_ptr<Room>& room,
+                     const std::shared_ptr<Doctor>& doctor)
+    : ymd_{ymd}, room_{room}, doctor_{doctor} {};
```
Comment (fbtom) [inline src/app/calendar/connector.cpp:5]: Shall we consider situation where we do not know yet if doctor will be available on the date, but we can reserve the room on that day and time?

File: src/app/clinic/clinic_facade.hpp [hunk 6]
```diff
@@ -15,21 +16,25 @@ class Clinic
     inline static std::vector<std::shared_ptr<Visit>> visits_;
     inline static std::vector<std::shared_ptr<Receptionist>> receptionists_;
     inline static std::vector<std::shared_ptr<Room>> rooms_;
+    inline static std::shared_ptr<Calendar> calendar_;
```
Comment (fbtom) [inline src/app/clinic/clinic_facade.hpp:19]: **idea**
As we spotted some issues with the static containers I suggest that in the future we will create an object Clinic which will take care of lifetime of other objects.
In such term, once Clinic will be created in main, we will have only one calendar initialized and then used by dependency injection. 
And then, if there will be single Calendar&, then we can remove checking if it's the only Calendar. There would not be a possibility to create another :D

**Review by fbtom — COMMENTED** [review]
General comments, just removing std::shared_ptr within calendar is requested ;)

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

File: src/app/calendar/CMakeLists.txt (hunk 1 above)
Comment (BaqWin) [reply src/app/calendar/CMakeLists.txt:3]: Added

File: src/app/calendar/calendar.cpp (hunk 2 above)
Comment (BaqWin) [reply src/app/calendar/calendar.cpp]: Added

File: src/app/clinic/clinic_facade.hpp (hunk 3 above)
Comment (BaqWin) [reply src/app/clinic/clinic_facade.hpp]: Added

File: src/app/calendar/calendar.hpp (hunk 4 above)
Comment (BaqWin) [reply src/app/calendar/calendar.hpp]: Changed

File: src/app/calendar/connector.cpp (hunk 5 above)
Comment (BaqWin) [reply src/app/calendar/connector.cpp:5]: I think there would be no such situation, why would we store information on a visit to a particular room if we don't know which doctor is going to be taking these visits?

File: src/app/clinic/clinic_facade.hpp (hunk 6 above)
Comment (BaqWin) [reply src/app/clinic/clinic_facade.hpp:19]: Yes, that's one way to do it, but what if we want to expand our business and make a new branch? We can consider adding support for multiple calendars with distinction to which branch it is from.

File: src/app/calendar/connector.cpp (hunk 5 above)
Comment (fbtom) [reply src/app/calendar/connector.cpp:5]: Refer to reserve queue, solved

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

**Review by fbtom — APPROVED** [review]

**Comment by ziobron** [comment]
Thanks @fbtom for Code Review.
🏅 1 XP granted

**Comment by ziobron** [comment]
Your PR was merged!
🏅 2 XP granted. Thanks @BaqWin for making course materials better!

**Pull request merged**